Fixed quantile nan propagation and implemented nanquantile

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44497 (nan)quantile CPU optimization
* **#44393 Fixed quantile nan propagation and implemented nanquantile**

torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

heitorschueroff committed:
Fixed quantile nan propagation and implemented nanquantile

[ghstack-poisoned]

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4610,11 +4610,11 @@ def merge_dicts(*dicts):
-beign flattened before computation.
+being flattened before computation.
 
 If :attr:`q` is a 1D tensor, the first dimension of the result corresponds to the quantiles
```
"corresponds to the quantiles" and "what remains from the reduction" are vague expressions. Can we be more precise here?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
+           r"""
+nanquantile(input, q, dim=None, keepdim=False, *, out=None) -> Tensor
+
+This is a variant of :func:`torch.quantile` that ignores `NaN` values when
```
... that "ignores" NaN values, computing the quantiles :attr:\`q\` as if NaN values in :attr:\`input\` did not exist. See the documentation for :func:\`torch.quantile\` and the example below.

I don't think you need to mention when NaN is produced in the output, but if you do I think it's more consistent to use terminology like "if all values in a reduced row are NaN..."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
+    >>> a.nanquantile(0.5, 1)
+    tensor([   nan, 1.5000])
+
+See :func:`torch.quantile`
```
Kill this line here. Typically we don't put anything after the example(s)

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
+    >>> torch.tensor([float('nan'), 1, 2]).nanquantile(0.5)
+    tensor(1.5000)
+
+    >>> a = torch.tensor([[float('nan'), float('nan')], [1, 2]])
```
nit: t =

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
+    >>> a
+    tensor([[nan, nan],
+            [1., 2.]])
+    >>> a.nanquantile(0.5, 0)
```
nit: add the dim= explicitly for clarity

mruberry reviewed (approved):
Looks good,

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed quantile nan propagation and implemented nanquantile"


torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

[ghstack-poisoned]
